### PR TITLE
Implement getCatalogSessionProperties

### PR DESCRIPTION
**Pull request by rclaude**
What: Add support for setting default catalog session properties in FileSessionPropertyManager & DbSessionPropertyManager
How: Use the convention that the first dot in a property name is a separator between a catalog name and a the catalog specific property name

This PR is a replacement for https://github.com/prestosql/presto/pull/5188

### CODE REVIEW

File: presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import io.prestosql.spi.session.SessionPropertyConfigurationManager;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public abstract class AbstractSessionPropertyManager
+        implements SessionPropertyConfigurationManager
```
Comment (kokosing) [inline presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java:27]: Please extract `AbstractSessionPropertyManager` extraction as separate commit. So logical changes are separated from refactors. Now it is more difficult to track how it changed.

File: presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    @Override
+    public Map<String, String> getSystemSessionProperties(SessionConfigurationContext context)
+    {
+        Map<String, String> combinedSystemSessionProperties = getSessionProperties(context)
+                .entrySet()
+                .stream()
+                .filter(entry -> !entry.getKey().contains("."))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
```
Comment (kokosing) [inline presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java:48]: collect to immutable map directly

Comment (rclaude) [reply presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java]: Done
Thanks! didn't knew about ImmutableMap collectors

File: presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java (hunk 1 above)
Comment (rclaude) [reply presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java:26]: Cut the change in 4 commits to ease the review

Comment (kokosing) [reply presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java:26]: Thanks!

File: presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java [hunk 3]
```diff
@@ -26,15 +26,24 @@
...
+        return combinedProperties;
+    }
+
+    @Override
+    public Map<String, String> getSystemSessionProperties(SessionConfigurationContext context)
+    {
+        return getSessionProperties(context)
+                .entrySet()
+                .stream()
+                .filter(entry -> !entry.getKey().contains("."))
```
Comment (kokosing) [inline presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java:46]: extract this condition as method: `!isCatalogSessionProperty()`

File: presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java [hunk 4]
```diff
@@ -26,15 +26,24 @@
...
+                .collect(ImmutableMap.toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
```
Comment (kokosing) [inline presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java:47]: static import

File: presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java [hunk 5]
```diff
@@ -49,7 +50,22 @@
     @Override
     public Map<String, Map<String, String>> getCatalogSessionProperties(SessionConfigurationContext context)
     {
-        // NOT IMPLEMENTED YET
-        return ImmutableMap.of();
+        Map<String, Map<String, String>> catalogsSessionProperties = new HashMap<>();
```
Comment (kokosing) [inline presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java:53]: Use ImmutableMap.Builder

**Review by kokosing — APPROVED** [review]

File: presto-session-property-managers/src/main/java/io/prestosql/plugin/session/file/FileSessionPropertyManager.java [hunk 6]
```diff
@@ -15,36 +15,33 @@
...
 public class FileSessionPropertyManager
-        implements SessionPropertyConfigurationManager
+        extends AbstractSessionPropertyManager
 {
     public static final JsonCodec<List<SessionMatchSpec>> CODEC = new JsonCodecFactory(
             () -> new ObjectMapperProvider().get().enable(FAIL_ON_UNKNOWN_PROPERTIES))
             .listJsonCodec(SessionMatchSpec.class);
 
-    private final List<SessionMatchSpec> sessionMatchSpecs;
+    private final ImmutableList<SessionMatchSpec> sessionMatchSpecs;
```
Comment (phd3) [inline presto-session-property-managers/src/main/java/io/prestosql/plugin/session/file/FileSessionPropertyManager.java:44]: We could just use interface type in the declaration

File: presto-docs/src/main/sphinx/admin/session-property-managers.rst [hunk 7]
```diff
@@ -60,7 +60,7 @@ Consider the following set of requirements:
   limit of 1 hour (tighter than the constraint on ``global``).
 
 * All ETL queries (tagged with 'etl') are routed to subgroups under the ``global.pipeline`` group, and must be
-  configured with certain properties to control writer behavior.
+  configured with certain properties to control writer behavior and a hive catalog property.
```
Comment (phd3) [inline presto-docs/src/main/sphinx/admin/session-property-managers.rst:63]: It'll be good to modify the `Limitations` section in this file to be consistent with the new behavior.

File: presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+import java.util.List;
+import java.util.Map;
+
+public abstract class AbstractSessionPropertyManager
+        implements SessionPropertyConfigurationManager
+{
+    protected abstract List<SessionMatchSpec> getSessionMatchSpecs();
+
+    @Override
+    public Map<String, String> getSystemSessionProperties(SessionConfigurationContext context)
```
Comment (phd3) [inline presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java]: I think we may want to `getSystemSessionProperties` and `getCatalogSessionProperties` final in this class to future-proof against possible subclasses that may try to override them. @kokosing what do you think?

File: presto-session-property-managers/src/test/java/io/prestosql/plugin/session/db/TestDbSessionPropertyManager.java [hunk 9]
```diff
@@ -82,14 +83,15 @@ public void destroy()
     }
 
     @Override
-    protected void assertProperties(Map<String, String> properties, SessionMatchSpec... specs)
+        protected void assertProperties(Map<String, String> systemProperties, Map<String, Map<String, String>> catalogProperties, SessionMatchSpec... specs)
```
Comment (phd3) [inline presto-session-property-managers/src/test/java/io/prestosql/plugin/session/db/TestDbSessionPropertyManager.java:86]: nit: indent

**Review by phd3 — COMMENTED** [review]
Thanks @rclaude, this looks good. Just a few minor comments.

File: presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java (hunk 3 above)
Comment (rclaude) [reply presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java]: Done

File: presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java (hunk 4 above)
Comment (rclaude) [reply presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java]: Done

File: presto-docs/src/main/sphinx/admin/session-property-managers.rst (hunk 7 above)
Comment (rclaude) [reply presto-docs/src/main/sphinx/admin/session-property-managers.rst:63]: Done

File: presto-session-property-managers/src/main/java/io/prestosql/plugin/session/file/FileSessionPropertyManager.java (hunk 6 above)
Comment (rclaude) [reply presto-session-property-managers/src/main/java/io/prestosql/plugin/session/file/FileSessionPropertyManager.java:44]: Made it explicit in the type so we can rely on it in getSessionMatchSpecs() and return the list without doing a ImmutableList.copyOf().

File: presto-session-property-managers/src/test/java/io/prestosql/plugin/session/db/TestDbSessionPropertyManager.java (hunk 9 above)
Comment (rclaude) [reply presto-session-property-managers/src/test/java/io/prestosql/plugin/session/db/TestDbSessionPropertyManager.java]: Fixed

File: presto-session-property-managers/src/main/java/io/prestosql/plugin/session/file/FileSessionPropertyManager.java (hunk 6 above)
Comment (rclaude) [reply presto-session-property-managers/src/main/java/io/prestosql/plugin/session/file/FileSessionPropertyManager.java:44]: Made it explicit in the type so we can rely on it in getSessionMatchSpecs() and return the list without doing a ImmutableList.copyOf().

File: presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java (hunk 8 above)
Comment (rclaude) [reply presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java]: Done

File: presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java (hunk 5 above)
Comment (rclaude) [reply presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java]: Partially done
Tried using groupingby collector or a custom collector to do it in a single path but only came up with more verbose & harder to grasp implementations.

File: presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java [hunk 10]
```diff
@@ -20,21 +20,38 @@
...
         // later properties override earlier properties
         Map<String, String> combinedProperties = new HashMap<>();
         for (SessionMatchSpec sessionMatchSpec : getSessionMatchSpecs()) {
             combinedProperties.putAll(sessionMatchSpec.match(context));
         }
-        return ImmutableMap.copyOf(combinedProperties);
+        return combinedProperties;
+    }
+
+    private static boolean isCatalogSessionProperty(Entry<String, String> property)
```
Comment (kokosing) [inline presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java:43]: move it to the bottom, is less interesting than public API

**Review by kokosing — APPROVED** [review]

File: presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java [hunk 11]
```diff
@@ -20,21 +20,38 @@
...
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
 
 public abstract class AbstractSessionPropertyManager
         implements SessionPropertyConfigurationManager
 {
     protected abstract List<SessionMatchSpec> getSessionMatchSpecs();
 
-    @Override
-    public final Map<String, String> getSystemSessionProperties(SessionConfigurationContext context)
+    private Map<String, String> getSessionProperties(SessionConfigurationContext context)
```
Comment (kokosing) [inline presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java:33]: put private methods below public. Put method definitions under their usage.

File: presto-session-property-managers/src/test/java/io/prestosql/plugin/session/db/TestDbSessionPropertyManager.java [hunk 12]
```diff
@@ -211,4 +213,25 @@ public void testOrderingOfSpecs()
...
+    @Test
+    public void testCatalogSessionProperties()
+    {
+        dao.insertSpecRow(1, ".*", null, null, null, 0);
+        dao.insertSessionProperty(1, "catalog_1.prop_1", "val_1");
+        dao.insertSessionProperty(1, "catalog_1.prop_2", "val_2");
+
+        dao.insertSpecRow(2, ".*", null, null, null, 1);
+        dao.insertSessionProperty(2, "catalog_1.prop_1", "val_1_bis");
+        dao.insertSessionProperty(2, "catalog_1.prop_3", "val_3");
```
Comment (kokosing) [inline presto-session-property-managers/src/test/java/io/prestosql/plugin/session/db/TestDbSessionPropertyManager.java:229]: can you please add more tests. Where are two catalogs, where the are no. When the are some system session properties. Where input is empty etc.

Comment (rclaude) [reply presto-session-property-managers/src/test/java/io/prestosql/plugin/session/db/TestDbSessionPropertyManager.java:229]: Added test for empty tables, empty strings and dots handling edge cases. 
For 2 catalogs, no catalog, those are covered by the shared tests in AbstractTestSessionPropertyManager.

File: presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java (hunk 10 above)
Comment (rclaude) [reply presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java:78]: Done

File: presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java (hunk 11 above)
Comment (rclaude) [reply presto-session-property-managers/src/main/java/io/prestosql/plugin/session/AbstractSessionPropertyManager.java:68]: Done

**Comment by rclaude** [comment]
I tried to rebase but still face ci errors that are unrelated with the PR, don't know how to go about that.
This time it's:

`tests               | 2020-10-06 13:19:24 INFO: Testing with bucketingType=BUCKETED_V1, value='prestosql rocks', insertWithPresto=false, expectedFileNamePossibilites=[000001_0]
tests               | 2020-10-06 13:19:25 INFO: FAILURE     /    io.prestosql.tests.hive.TestHiveBucketedTables.testBucketingVersion (Groups: ) took 49.2 seconds
tests               | 2020-10-06 13:19:25 SEVERE: Failure cause:
tests               | io.prestosql.tempto.query.QueryExecutionException: java.sql.SQLException: Error while processing statement: FAILED: Execution Error, return code 1 from org.apache.hadoop.hive.ql.exec.mr.MapRedTask. Error caching map.xml: java.nio.channels.ClosedByInterruptException
tests               | 	at io.prestosql.tempto.query.JdbcQueryExecutor.execute(JdbcQueryExecutor.java:114)
tests               | 	at io.prestosql.tempto.query.JdbcQueryExecutor.executeQuery(JdbcQueryExecutor.java:82)
tests               | 	at io.prestosql.tests.hive.TestHiveBucketedTables.testBucketingVersion(TestHiveBucketedTables.java:304)
tests               | 	at io.prestosql.tests.hive.TestHiveBucketedTables.testBucketingVersion(TestHiveBucketedTables.java:280)
tests               | 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
tests               | 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
tests               | 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
tests               | 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
tests               | 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
tests               | 	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
tests               | 	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
tests               | 	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
tests               | 	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
tests               | 	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
tests               | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
tests               | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
tests               | 	at java.base/java.lang.Thread.run(Thread.java:834)
tests               | Caused by: java.sql.SQLException: Error while processing statement: FAILED: Execution Error, return code 1 from org.apache.hadoop.hive.ql.exec.mr.MapRedTask. Error caching map.xml: java.nio.channels.ClosedByInterruptException
tests               | 	at org.apache.hive.jdbc.HiveStatement.execute(HiveStatement.java:275)
tests               | 	at io.prestosql.tempto.query.JdbcQueryExecutor.executeQueryNoParams(JdbcQueryExecutor.java:122)
tests               | 	at io.prestosql.tempto.query.JdbcQueryExecutor.execute(JdbcQueryExecutor.java:107)
tests               | 	... 16 more`

File: presto-session-property-managers/src/test/java/io/prestosql/plugin/session/db/TestDbSessionPropertyManager.java [hunk 13]
```diff
@@ -211,4 +213,37 @@ public void testOrderingOfSpecs()
         assertEquals(sessionProperties.get("prop_3"), "val_3_1");
         assertEquals(sessionProperties.size(), 3);
     }
+
+    /**
+     * A basic test for catalog session properties overrides with the {@link DbSessionPropertyManager}
+     */
```
Comment (kokosing) [inline presto-session-property-managers/src/test/java/io/prestosql/plugin/session/db/TestDbSessionPropertyManager.java:219]: This comment is not needed.

File: presto-session-property-managers/src/test/java/io/prestosql/plugin/session/db/TestDbSessionPropertyManager.java [hunk 14]
```diff
@@ -211,4 +213,37 @@ public void testOrderingOfSpecs()
...
+        specsProvider.refresh();
+        SessionConfigurationContext context1 = new SessionConfigurationContext("foo", Optional.empty(), ImmutableSet.of(), Optional.empty(), TEST_RG);
+        assertEquals(manager.getCatalogSessionProperties(context1),
+                ImmutableMap.of("catalog_1",
+                        ImmutableMap.of("prop_1", "val_1_bis", "prop_2", "val_2", "prop_3", "val_3")));
+    }
+
+    /**
+     * Test empty tables
+     */
```
Comment (kokosing) [inline presto-session-property-managers/src/test/java/io/prestosql/plugin/session/db/TestDbSessionPropertyManager.java:240]: This comment is not needed.

**Review by kokosing — APPROVED** [review]
Please rebase and apply last comments. It looks like you hit https://github.com/prestosql/presto/issues/5427, hopefully after the rebase the issue won't appear again.

**Comment by rclaude** [comment]
> Please rebase and apply last comments. It looks like you hit #5427, hopefully after the rebase the issue won't appear again.

Got a green CI !

**Comment by kokosing** [comment]
Merged, thanks!
